src/main/java/org/opensearch/neuralsearch/query/HybridQueryScorer.java

```diff
@@ -0,0 +1,144 @@
+import java.util.Map;
+
+import lombok.Getter;
+
+import org.apache.lucene.search.DisiPriorityQueue;
+import org.apache.lucene.search.DisiWrapper;
+import org.apache.lucene.search.DisjunctionDISIApproximation;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.Weight;
+
+/**
+ * Class abstracts functionality of Scorer for hybrid query. When iterating over documents in increasing
+ * order of doc id, this class fills up array of scores per sub-query for each doc id. Order in array of scores
+ * corresponds to order of sub-queries in an input Hybrid query.
+ */
+public final class HybridQueryScorer extends Scorer {
+
+    // score for each of sub-query in this hybrid query
+    @Getter
+    private final Scorer[] subScorers;
+
+    private final DisiPriorityQueue subScorersPQ;
+
+    private final DocIdSetIterator docIdIterator;
+
+    final private float[] subScores;
+
+    final private Map<Query, Integer> queryToIndex;
+
+    HybridQueryScorer(Weight weight, Scorer[] subScorers) throws IOException {
+        super(weight);
+        this.subScorers = subScorers;
+        queryToIndex = new HashMap<>();
+        subScores = new float[subScorers.length];
+        int idx = 0;
+        int size = 0;
+        for (Scorer scorer : subScorers) {
+            if (scorer == null) {
+                idx++;
+                continue;
+            }
+            queryToIndex.put(scorer.getWeight().getQuery(), idx);
+            idx++;
+            size++;
+        }
+        this.subScorersPQ = new DisiPriorityQueue(size);
+        for (Scorer scorer : subScorers) {
+            if (scorer == null) {
+                continue;
+            }
+            final DisiWrapper w = new DisiWrapper(scorer);
+            this.subScorersPQ.add(w);
+        }
+        this.docIdIterator = new DisjunctionDISIApproximation(this.subScorersPQ);
+    }
+
+    /**
+     * Returns the score of the current document matching the query. Score is a sum of all scores from sub-query scorers.
+     * @return combined total score of all sub-scores
+     * @throws IOException
+     */
+    @Override
+    public float score() throws IOException {
+        DisiWrapper topList = subScorersPQ.topList();
+        float totalScore = 0.0f;
+        for (DisiWrapper disiWrapper = topList; disiWrapper != null; disiWrapper = disiWrapper.next) {
+            // check if this doc has match in the subQuery. If not, add score as 0.0 and continue
+            if (disiWrapper.scorer.docID() == DocIdSetIterator.NO_MORE_DOCS) {
+                continue;
+            }
+            totalScore += disiWrapper.scorer.score();
+        }
+        return totalScore;
+    }
+
+    /**
+     * Return a DocIdSetIterator over matching documents.
+     * @return DocIdSetIterator object
+     */
+    @Override
+    public DocIdSetIterator iterator() {
+        return docIdIterator;
```

Comment:
should't we return a copy of iterator? otherwise, it can be used only one time.

Reply:
This would be nice but there is no existing method to clone iterator. So far I didn't notice any problems with a single iterator instance, I suggest we keep it as it is for now and address this later

Reply:
you can do `return new DisjunctionDISIApproximation(this.subScorersPQ);`?

Reply:
I'm not sure if that equivalent code, every time we'll get new instance, and with current implementation it's same instance that saves state. Let me try if that cause any test to fail.